### ISSUE TITLE
Codelens is not working with code that does import turtle

### ISSUE THREAD

**Issue by richard-delorenzi** [RunestoneInteractive/RunestoneComponents#524]
If I use activecode with codelens, or codelens, and I have a code line `import turtle`, then when the import line is executed the following error is produced.
`ImportError: turtle not supported`

----

I am using version 3.0.7 installed on a barebones debian jessie, in docker.
see [Dockerfile](https://bitbucket.org/mr_delorenzi_teacher/python-from-scratch/src/27bcca6c4db6f21244b284b34cad0335317f2bfe/docker/builder/Dockerfile?at=default&fileviewer=file-view-default)

```
FROM debian:jessie

#########################################################
#Proxy

RUN apt-get update
RUN apt-get install -y libproxy-tools

ENV http_proxy 'http://172.17.0.1:3128'
ENV https_proxy 'http://172.17.0.1:3128'
ADD aptconf-autoproxy /etc/apt/apt.conf.d/30autoproxy

################################################################
# install pip installer

RUN apt-get update && apt-get install -y  python-pip

################################################################
# install runestone

RUN pip install runestone

################################################################

# Replace 1000 with your user / group id
RUN export uid=1000 gid=1000 && \
    mkdir -p /home/developer && \
    echo "R.Delorenzi:x:${uid}:${gid}:R.Delorenzi,,,:/home/developer:/bin/bash" >> /etc/passwd && \
    echo "R.Delorenzi:x:${uid}:" >> /etc/group && \
    chown ${uid}:${gid} -R /home/developer

USER R.Delorenzi
ENV HOME /home/developer
WORKDIR /home/developer

################################################################
#bash

################################################################
#

LABEL name="runestone bash" \
      version="" \
      description="a tool for making interactive programming books"
```

**Comment by bnmnetp** [comment]
Yes, that is the expected behavior.

Activecode is run by an implementation of Python in the browser.  CodeLens is run on the pythontutor.com server in a sandboxed environment, where we can't support interactive graphics.

**Comment by richard-delorenzi** [comment]
Are you sure that it runs at pythontutor.com? I can build off-line.

**Comment by richard-delorenzi** [comment]
Would it be possible to get it to work. Maybe with a dummy turtle package, when calculating the codelens stuff. But keeping the real turtle for the python in browser.

This is the sort of coding trick that I have used in the past, on complex projects.
The dummy package would not have much: e.g there is a turtle.forward(n), but nothing about what it does. The harder part is to get the builder to pass one to the codelens builder, and the other to the python in browser.

It would not have to support the whole turtle library. Just the subset that is needed by a new student (so long as the documentation was clear on this).

Hope this makes sense. If not leave a comment.

**Comment by bnmnetp** [comment]
When you build the book it uses a local version of Python using the modules from pythontutor.com.

When you press the show in codelens button it opens an iframe using the API of the live pythontutor.com

I'm sure what you suggest regarding getting the turtle to work is possible.  Its all a matter of time and priorities.  PR's are always welcome 😄

**Comment by EricGT** [comment]
@richard-delorenzi 

I don't know the level of the students you are teaching but some reflection from what I have found with regards to your question that may be of help to you and others.

The reason I don't see a lot of people trying to pursue getting CodeLens to work with Turtle Graphics is that it is going to take some coding to get it to work, possibly significant. So while doable I suspect that you will find that if you teach the course as laid out that by the time the students reach the Turtle Graphics in Chapter 4 that they won't be needing to use CodeLens for feedback as the Turtle and the image it draws are typically more than enough feedback to understand how the code works and if anything where exactly the code went wrong. In other words you can think of CodeLens as a mini debugger that shows you the values, but that the Turtle is a different kind of debugger in that it leaves a visible pattern of what it does. In part I would suspect that is why grammar and middle school students take such a liking to learning to program with Turtle graphics is because the feedback is more intuitive than learning to understand how work with a debugger, or in this case CodeLens.

So while I think you have a noble idea, that maybe waiting to see if it is needed by the students would be better as your time and effort could be spent on something different.  In the real world of problem solving we don't always have what we want but in further searching this can sometimes lead to serendipitous answers that are more useful.

**Comment by richard-delorenzi** [comment]
I am starting turtle on day one. However I think you are correct. They may
not need the codelens. I changed the way that I wrote my book, and will see
how it goes.  If I think it is useful, then I will look at your code and
make an estimate of the complexity to add this feature. If low I will do
it. But I will have to wait until a holiday (tomorrow is back to teaching).

Thanks,
Mr R.Delorenzi

--
This e-mail was sent from a computer running Debian <http://www.debian.org/>,
a Free <http://en.wikipedia.org/wiki/Free_software> Gnu/Linux operating
system.
